### PR TITLE
[Plugin-1703] Support chinese column names

### PR DESCRIPTION
**Pull request by psainics**
## Support for Chinese character in column names [ Bug Fix ]

Jira : [Plugin-1703](https://cdap.atlassian.net/browse/PLUGIN-1703)

### Changes
- Enabled extended character support for field/column names.
- As per docs max column name can now be 300
- New regex for valid and invalid field/column names.
- Update function `isJsonFormat` to return true when non English character is used
   -  We do this as the avro load job in BQ does not support non-english characters in field names for now
   
### UI Field
- A new ui field is added (false by default) to maintain backwards compatibility, where users can trigger old mappings. 
image

### Unit Test 
- Ran old unit test as validation
image

- Extended test cases with
  - Contains hyphen 
  - Contains Chinese character 
  - Contains Japanese character
  - Contains Emoji

### CODE REVIEW

File: src/main/java/io/cdap/delta/bigquery/BigQueryUtils.java [hunk 1]
```diff
@@ -68,6 +68,8 @@ public final class BigQueryUtils {
   // Pc (connector, including underscore), Pd (dash), Zs (space).
   // See here: https://cloud.google.com/bigquery/docs/tables#table_naming
   private static final Pattern VALID_TABLE_NAME_REGEX = Pattern.compile("[\\p{L}\\p{M}\\p{N}\\p{Pc}\\p{Pd}\\p{Zs}]+");
+  private static final Pattern VALID_FIELD_NAME_REGEX = Pattern.compile("[\\p{L}\\p{M}\\p{N}\\p{Pc}\\p{Pd}]+");
```
Comment (MrRahulSharma) [inline src/main/java/io/cdap/delta/bigquery/BigQueryUtils.java]: Can you confirm if space is not allowed ? As per the documentation here : https://cloud.google.com/bigquery/docs/schemas#flexible-column-names, it conflicts.

Comment (psainics) [reply src/main/java/io/cdap/delta/bigquery/BigQueryUtils.java]: Updated the PR with new requirements.

File: src/test/java/io/cdap/delta/bigquery/BigQueryUtilsTest.java [hunk 2]
```diff
@@ -153,23 +153,34 @@ public void testNormalizeTableName() {
...
-      name = Strings.repeat("a1", 65);
-      assertEquals(name.substring(0, 128), BigQueryUtils.normalizeFieldName(name));
+      name = Strings.repeat("a1", 151);
+      assertEquals(name.substring(0, maxColumnNameLength), BigQueryUtils.normalizeFieldName(name, false));
       // contains invalid character
-      assertEquals("ab_c", BigQueryUtils.normalizeFieldName("ab?/c"));
+      assertEquals("ab_c", BigQueryUtils.normalizeFieldName("ab?/c", false));
       // contains space
-      assertEquals("a2_fs", BigQueryUtils.normalizeFieldName("a2 fs"));
+      assertEquals("a2_fs", BigQueryUtils.normalizeFieldName("a2 fs", false));
```
Comment (MrRahulSharma) [inline src/test/java/io/cdap/delta/bigquery/BigQueryUtilsTest.java:175]: Please add a test for whitespace comparison as well.

File: docs/bigquery-cdcTarget.md [hunk 3]
```diff
@@ -82,6 +82,12 @@ name is same as source database name. A valid name should only contain letters,
 maximum length can be 1024. Any invalid chars would be replaced with underscore in the final dataset name and
 any characters exceeds length limit will be truncated.
 
+**Allow Flexible Column Naming**:
+By default ony english letters, numbers and underscore are allowed in column names. If this option is enabled,
+international characters are also allowed in column names with some extra special characters, which follow the
```
Comment (MrRahulSharma) [inline docs/bigquery-cdcTarget.md:87]: It would help if we can list a few chars which will get supported by this change.

File: src/main/java/io/cdap/delta/bigquery/BigQueryUtils.java (hunk 1 above)
Comment (MrRahulSharma) [reply src/main/java/io/cdap/delta/bigquery/BigQueryUtils.java]: What is the resolution here, are we supporting whitespaces in the field names ?

File: docs/bigquery-cdcTarget.md (hunk 3 above)
Comment (psainics) [reply docs/bigquery-cdcTarget.md:87]: Updated docs, incuded
> Some special characters allowed in flexible column names are: 
> - An ampersand (&)
> - A percent sign (%)
> - A colon (:)
> - A lessthan sign (<)
> - A space ( )

File: src/test/java/io/cdap/delta/bigquery/BigQueryUtilsTest.java (hunk 2 above)
Comment (psainics) [reply src/test/java/io/cdap/delta/bigquery/BigQueryUtilsTest.java:175]: Great! I've added tests for whitespace and covered all the valid and invalid characters specified in the documentation.

File: src/main/java/io/cdap/delta/bigquery/BigQueryUtils.java (hunk 1 above)
Comment (psainics) [reply src/main/java/io/cdap/delta/bigquery/BigQueryUtils.java]: Yes, we are supporting all valid flexible names, with whitespaces and other special characters.
This is the new regex `"[\\p{L}\\p{M}\\p{N}\\p{Pc}\\p{Pd}&%+=:'<>#|\\s]+"`

File: docs/bigquery-cdcTarget.md [hunk 4]
```diff
@@ -82,6 +82,19 @@ name is same as source database name. A valid name should only contain letters,
...
+bigquery naming convention for flexible column names.
+Some special characters allowed in flexible column names are: 
+- An ampersand (&)
+- A percent sign (%)
+- A colon (:)
+- A lessthan sign (<)
+- A space ( )
```
Comment (MrRahulSharma) [inline docs/bigquery-cdcTarget.md:94]:
```suggestion
- Ampersand (&)
- Percent (%)
- Colon (:)
- Less than (<)
- Whitespace ( )
```

**Review by MrRahulSharma — APPROVED** [review]
